### PR TITLE
TEST: TMA-1040 Add tests for "perform" method in LCM2 module

### PR DESCRIPTION
**Pull request by sadam21**

### CODE REVIEW

File: spec/unit/lcm/lcm2_spec.rb [hunk 1]
```diff
@@ -45,10 +37,125 @@
...
+      let(:params) do
+        params = {
+          GDC_GD_CLIENT: client,
+          GDC_LOGGER: logger
+        }
+        expect { raise RuntimeError }.to raise_error
+        GoodData::LCM2.convert_to_smart_hash(params)
+      end
+
+      it 'exception should be thrown' do
```
Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: warning: Similar code found in lcm2_spec.rb:68, lcm2_spec.rb:103 (mass = 40) (flay)

File: spec/unit/lcm/lcm2_spec.rb [hunk 2]
```diff
@@ -45,10 +37,125 @@
...
+
+    context 'when invalid mode is specified' do
+      let(:params) do
+        params = {
+          GDC_LOGGER: logger
+        }
+        GoodData::LCM2.convert_to_smart_hash(params)
+      end
+
+      it 'should fail' do
```
Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: warning: Similar code found in lcm2_spec.rb:68, lcm2_spec.rb:103 (mass = 40) (flay)

File: spec/unit/lcm/lcm2_spec.rb [hunk 3]
```diff
@@ -45,10 +37,125 @@
...
+            }
+            GoodData::LCM2.convert_to_smart_hash(params)
+          end
+
+          class Action1 < GoodData::LCM2::BaseAction
+            PARAMS = { }
+            DESCRIPTION = "Test action"
+          end
+
+          class Action2 < Action1
```
Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: info: Has no descriptive comment (IrresponsibleModule) (reek)

Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: info: Has the name 'action2' (UncommunicativeModuleName) (reek)

File: spec/unit/lcm/lcm2_spec.rb [hunk 4]
```diff
@@ -45,10 +37,125 @@
...
+          end
+
+          class Action3 < Action1
```
Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: info: Has no descriptive comment (IrresponsibleModule) (reek)

Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: info: Has the name 'action3' (UncommunicativeModuleName) (reek)

File: spec/unit/lcm/lcm2_spec.rb [hunk 5]
```diff
@@ -45,10 +37,125 @@
...
+          let(:params) do
+            params = {
+              GDC_LOGGER: logger,
+              fail_early: fail_early,
+              strict: strict_mode
+            }
+            GoodData::LCM2.convert_to_smart_hash(params)
+          end
+
+          class Action1 < GoodData::LCM2::BaseAction
```
Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: info: Has the name 'action1' (UncommunicativeModuleName) (reek)

File: spec/unit/lcm/lcm2_spec.rb [hunk 6]
```diff
@@ -45,10 +37,125 @@
...
+          GDC_LOGGER: logger
+        }
+        expect { raise RuntimeError }.to raise_error
+        GoodData::LCM2.convert_to_smart_hash(params)
+      end
+
+      it 'exception should be thrown' do
+        expect do
+          GoodData::LCM2.perform('users', params)
+        end.to raise_error { |e| expect(e.message).to include('Either organisation or domain has to be specified in params')}
```
Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: warning: Style/SpaceInsideBlockBraces: Space missing inside }. (rubocop)

File: spec/unit/lcm/lcm2_spec.rb [hunk 7]
```diff
@@ -45,10 +37,125 @@
...
+        params = {
+          GDC_LOGGER: logger
+        }
+        GoodData::LCM2.convert_to_smart_hash(params)
+      end
+
+      it 'should fail' do
+        expect do
+          GoodData::LCM2.perform('invalid', params)
+        end.to raise_error { |e| expect(e.message).to include('Invalid mode specified \'invalid\'')}
```
Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: warning: Style/SpaceInsideBlockBraces: Space missing inside }. (rubocop)

File: spec/unit/lcm/lcm2_spec.rb [hunk 8]
```diff
@@ -45,10 +37,125 @@
...
+      end
+    end
+
+    [ true, false ].each do |fail_early|
```
Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: warning: Style/SpaceInsideBrackets: Space inside square brackets detected. (rubocop)

File: spec/unit/lcm/lcm2_spec.rb [hunk 9]
```diff
@@ -45,10 +37,125 @@
...
+      [ true, false ].each do |strict_mode|
```
Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: warning: Style/SpaceInsideBrackets: Space inside square brackets detected. (rubocop)

File: spec/unit/lcm/lcm2_spec.rb [hunk 10]
```diff
@@ -45,10 +37,125 @@
...
+            params = {
+              GDC_LOGGER: logger,
+              fail_early: fail_early,
+              strict: strict_mode
+            }
+            GoodData::LCM2.convert_to_smart_hash(params)
+          end
+
+          class Action1 < GoodData::LCM2::BaseAction
+            PARAMS = { }
```
Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: warning: Style/SpaceInsideHashLiteralBraces: Space inside empty hash literal braces detected. (rubocop)

File: spec/unit/lcm/lcm2_spec.rb [hunk 11]
```diff
@@ -45,10 +37,125 @@
...
+          end
+
+          class Action2 < Action1
+          end
+
+          class Action3 < Action1
+          end
+
+          before do
+            allow(GoodData::LCM2).to receive(:get_mode_actions) { [ Action1, Action2, Action3 ] }
```
Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: warning: Style/SpaceInsideBrackets: Space inside square brackets detected. (rubocop)

File: spec/unit/lcm/lcm2_spec.rb [hunk 12]
```diff
@@ -45,10 +37,125 @@
...
+            expect(Action1).to receive(:call).and_return([{:a1 => 'a1'}])
```
Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: warning: Style/SpaceInsideHashLiteralBraces: Space inside { missing. (rubocop)

Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: warning: Style/SpaceInsideHashLiteralBraces: Space inside } missing. (rubocop)

File: spec/unit/lcm/lcm2_spec.rb [hunk 13]
```diff
@@ -45,10 +37,125 @@
...
+            expect(Action2).to receive(:call).and_raise("boom")
+            if fail_early
+              expect(Action3).not_to receive(:call)
+            else
+              expect(Action3).to receive(:call).and_return([{:a3 => 'a3'}])
```
Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: warning: Style/SpaceInsideHashLiteralBraces: Space inside { missing. (rubocop)

Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: warning: Style/SpaceInsideHashLiteralBraces: Space inside } missing. (rubocop)

File: spec/unit/lcm/lcm2_spec.rb [hunk 14]
```diff
@@ -45,10 +37,125 @@
...
+            end
+          end
+
+          it "should fail and #{fail_early ? 'stop' : 'continue'} performing next actions" do
+            if strict_mode
+              expect do
+                GoodData::LCM2.perform('some_mode', params)
+              end.to raise_error { |e| expect(e.message).to include('boom')}
```
Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: warning: Style/SpaceInsideBlockBraces: Space missing inside }. (rubocop)

File: spec/unit/lcm/lcm2_spec.rb [hunk 15]
```diff
@@ -45,10 +37,125 @@
...
+            else
+              result = GoodData::LCM2.perform('some_mode', params)
+              expect(result[:actions]).to eq(['Action1', 'Action2', 'Action3'])
```
Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: warning: Style/WordArray: Use `%w` or `%W` for an array of words. (rubocop)

File: spec/unit/lcm/lcm2_spec.rb [hunk 16]
```diff
@@ -45,10 +37,125 @@
...
+              expect(result[:results]["Action1"]).to eq([{:a1 => 'a1'}])
```
Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: warning: Style/SpaceInsideHashLiteralBraces: Space inside { missing. (rubocop)

Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: warning: Style/SpaceInsideHashLiteralBraces: Space inside } missing. (rubocop)

File: spec/unit/lcm/lcm2_spec.rb [hunk 17]
```diff
@@ -45,10 +37,125 @@
...
+              expect(result[:results]["Action2"]).to eq(nil)
+              expect(result[:results]["Action3"]).to eq(fail_early ? nil : [{:a3 => 'a3'}])
```
Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: warning: Style/SpaceInsideHashLiteralBraces: Space inside { missing. (rubocop)

Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: warning: Style/SpaceInsideHashLiteralBraces: Space inside } missing. (rubocop)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/493/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1322/ : FAILURE in 2m 59s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1592/ : SUCCESS in 2m 44s

File: spec/unit/lcm/lcm2_spec.rb [hunk 18]
```diff
@@ -45,10 +37,128 @@
...
+            params = {
+              GDC_LOGGER: logger,
+              fail_early: fail_early,
+              strict: strict_mode
+            }
+            GoodData::LCM2.convert_to_smart_hash(params)
+          end
+
+          # Test class 1
+          class Action1 < GoodData::LCM2::BaseAction
```
Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: info: Has the name 'action1' (UncommunicativeModuleName) (reek)

File: spec/unit/lcm/lcm2_spec.rb [hunk 19]
```diff
@@ -45,10 +37,128 @@
...
+            PARAMS = {}
+            DESCRIPTION = "Test action"
+          end
+
+          # Test class 2
+          class Action2 < Action1
```
Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: info: Has the name 'action2' (UncommunicativeModuleName) (reek)

File: spec/unit/lcm/lcm2_spec.rb [hunk 20]
```diff
@@ -45,10 +37,128 @@
...
+          end
+
+          # Test class 3
+          class Action3 < Action1
```
Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: info: Has the name 'action3' (UncommunicativeModuleName) (reek)

File: spec/unit/lcm/lcm2_spec.rb [hunk 21]
```diff
@@ -45,10 +37,128 @@
...
+          end
+
+          before do
+            allow(GoodData::LCM2).to receive(:get_mode_actions) { [Action1, Action2, Action3] }
+            expect(Action1).to receive(:call).and_return([{:a1 => 'a1'}])
```
Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: warning: Style/SpaceInsideHashLiteralBraces: Space inside { missing. (rubocop)

Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: warning: Style/SpaceInsideHashLiteralBraces: Space inside } missing. (rubocop)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/495/ : SUCCESS in 6s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1324/ : FAILURE in 1m 22s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1594/ : SUCCESS in 2m 39s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/496/ : SUCCESS in 6s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1325/ : FAILURE in 1m 02s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1595/ : SUCCESS in 2m 43s

File: spec/unit/lcm/lcm2_spec.rb [hunk 22]
```diff
@@ -45,10 +37,111 @@
...
+    context 'when invalid mode is specified' do
+      let(:params) do
+        params = {
+          GDC_LOGGER: logger
+        }
+        GoodData::LCM2.convert_to_smart_hash(params)
+      end
+
+      it 'should throw exception' do
+        expect do
```
Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: warning: Similar code found in lcm2_spec.rb:87, lcm2_spec.rb:132 (mass = 34) (flay)

File: spec/unit/lcm/lcm2_spec.rb [hunk 23]
```diff
@@ -45,10 +37,111 @@
...
+            if fail_early
+              expect(Action3).not_to receive(:call)
+            else
+              expect(Action3).to receive(:call).and_return([{ :a3 => 'a3' }])
+            end
+          end
+
+          it "should fail and #{fail_early ? 'stop' : 'continue'} performing next actions" do
+            if strict_mode
+              expect do
```
Comment (yenkins) [inline spec/unit/lcm/lcm2_spec.rb]: warning: Similar code found in lcm2_spec.rb:87, lcm2_spec.rb:132 (mass = 34) (flay)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/497/ : SUCCESS in 27s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1326/ : FAILURE in 1m 12s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1596/ : SUCCESS in 1m 04s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/503/ : SUCCESS in 16s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1330/ : SUCCESS in 1m 08s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1600/ : SUCCESS in 1m 12s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/504/ : SUCCESS in 15s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1331/ : SUCCESS in 2m 48s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1601/ : SUCCESS in 2m 37s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-merge-compliance-check/115/ : SUCCESS in 26s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1332/ : SUCCESS in 1m 12s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1602/ : SUCCESS in 1m 12s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1490/ : SUCCESS in 12m 58s
- https://checklist.intgdc.com/job/lcm/job/gooddata-ruby-lcm-integration-tests-zuul/61/ : SUCCESS in 43m 17s

**Pull request merged**